include/boost/geometry/algorithms/detail/overlay/split_rings.hpp

```diff
@@ -0,0 +1,433 @@
+
+#include <boost/geometry/util/condition.hpp>
+
+#include <boost/geometry/algorithms/detail/signed_size_type.hpp>
+
+#include <boost/geometry/algorithms/detail/overlay/get_turn_info.hpp>
+#include <boost/geometry/algorithms/detail/overlay/overlay_type.hpp>
+#include <boost/geometry/algorithms/detail/overlay/self_turn_points.hpp>
+#include <boost/geometry/algorithms/detail/overlay/turn_info.hpp>
+
+
+namespace boost { namespace geometry
+{
+
+namespace detail { namespace overlay
+{
+
+
+// the ring_as_dcl class implements a ring using a doubly-connected list;
+// it is a model of a bidirectional-traversal Boost.Range and
+// supports:
+// (1) constant-time push_back(), size() and empty() operations;
+// (2) splitting at two vertices (in order to create two rings out of
+//     the initial ring) also in constant time.
+//
+// The second operation cannot be done in constant time using typical
+// random access ranges, which is precisely the reason why this class
+// has been introduced and used.
+template
+<
+    typename Point,
+    closure_selector Closure,
+    typename List = std::list<Point>
+>
+class ring_as_dcl
+{
+public:
+    typedef Point point_type;
+    typedef List list_type;
+    typedef typename List::size_type size_type;
+    typedef typename List::iterator iterator;
+    typedef typename List::const_iterator const_iterator;
+
+    ring_as_dcl()
+        : m_list()
+    {}
+
+    inline void push_back(Point const& point)
+    {
+        m_list.push_back(point);
+    }
+
+    inline void split_at(iterator pos1, iterator pos2, ring_as_dcl& other)
+    {
+        // for this method to work, I think that the iterator pos1
+        // must preceed iterator pos2 in the original list node sequence
+        other.m_list.splice(other.m_list.end(), m_list, pos1, pos2);
+
+        if (BOOST_GEOMETRY_CONDITION(Closure == closed))
+        {
+            other.m_list.push_back(*other.m_list.begin());
+        }
+    }
+
+    inline iterator begin() { return m_list.begin(); }
+    inline const_iterator begin() const { return m_list.begin(); }
+
+    inline iterator end() { return m_list.end(); }
+    inline const_iterator end() const { return m_list.end(); }
+
+    inline void swap(ring_as_dcl& other)
+    {
+        m_list.swap(other.m_list);
+    }
+private:
+    List m_list;
+};
+
+
+template <closure_selector Closure>
+class find_duplicate_points
+{
+    static bool const is_closed = Closure == closed;
+
+    struct point_iterator_less
+    {
+        template <typename PointIterator>
+        inline bool operator()(PointIterator it1, PointIterator it2) const
+        {
+            return geometry::less
+                <
+                    typename std::iterator_traits<PointIterator>::value_type
+                >()(*it1, *it2);
+        }
+    };
+
+public:
+    template <typename Ring, typename Iterator>
+    static inline bool apply(Ring& ring, Iterator& pos1, Iterator& pos2)
```

Comment:
Yes it should be `const` in the sense that the ring is not modified. On the other hand I need the `Iterator` to be a mutable iterator of the `Ring` type which means that internally I would have needed to convert a const iterator to a mutable iterator which is not possible.

What I did was to make `ring` a `Ring const&` and internally convert it to a non-const ring so that the iterators passed to `pos1` and `pos2` are of the correct type. See commit e542d0d.